### DISCUSSION TITLE
How to use C librpm API to query database and get package's header ?

### DISCUSSION THREAD

**Discussion by jerome-diver** [rpm-software-management/rpm#3322] [General]
I have rpm-dev and rpm version 4.19.1.1
After to read librpm API from [ftp API access version 4.19.0](http://ftp.rpm.org/api//4.19.0/)
And after to try to learn reading these links below:
- [Jeff's tuto](https://jfearn.fedorapeople.org/en-US/RPM/4/html/RPM_Guide/index.html) that seems to be a copy of official one
- [Oregon State University Open Source Lab examples](https://ftp.osuosl.org/pub/rpm/max-rpm/s1-rpm-rpmlib-example-code.html) i still don't find where it is indicate for which version of librpm to use with that code, but contains lot of unexisting function in version 4.19.1
- [official archive documentation](https://web.archive.org/web/20160927085905/https://docs.fedoraproject.org/en-US/Fedora_Draft_Documentation/0.1/html/RPM_Guide/ch15s04.html) where at some point, i'm supposed to know how to "/* Do something... */" inside a called iteration of headers...

I feel more and more confused by unrelevent examples or missing information or /*Do something... */ documentation text and so on.

Does it exist some quality documentation that can show nice examples on how to use this API library for many requested tasks whre you don't need to know the API prior to learn the API ?

I try to "do something" (as said in the official doc) with that code:

```C
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <rpm/rpmdb.h>
#include <rpm/rpmtd.h>
#include <rpm/rpmts.h>
#include <rpm/header.h>

void get_package_info(const char *package_name) {
  rpmts ts;                    // Selector to ask
  rpmdbMatchIterator iter;     // iterator for packages found
  Header h;                    // header to ask for tag typed values
  int itertag; // returned from founded header to be describe               
  // Create requested Selector to ask something to query
  ts = rpmtsCreate();
  // Initialize an iterator for querying packages to the database
  // We ask an iterator representing result about \
  // all the packages that have "rpm" as tag name.
  iter = rpmtsInitIterator(ts, RPMTAG_NAME, package_name, 0);
  // Iterate it to get each header returned
  while ((h = rpmdbNextIterator(iter)) != NULL) {
    HeaderIterator h_iter;
    // now start to iterate the header's tag for the package found
    h_iter = headerInitIterator(h);
    rpmtd td = rpmtdNew();
    while ( itertag = headerNext(h_iter, td)) {
      // And depend on the header found, apply to do something with data collect as:
      // "itertag" is the actual iterated tag to test
      switch (itertag) {
      case RPMTAG_VERSION:
        printf("Version: %s\n", rpmtdGetString(td));
        break;
      case RPMTAG_RELEASE:
        printf("Release: %s\n", rpmtdGetString(td));
        break;
      }
    }
  }

  // Cleanup
  rpmdbFreeIterator(iter);
  rpmtsFree(ts);
}
```
But get some error like:
- error: no dbpath has been set
- error: cannot open Packages database in

And after to add from example code i read above:
```C
  printf("Database path is: %s\n", rpmGetVar(RPMVAR_DBPATH) ? rpmGetVar(RPM_DBPATH) : "none");
```

sure, i got error about unexisting rpmGetVar function.

Well... from the command:
` rg -e "rpmGetVar" `
inside /usr/include/rpm directory, it return that nothing with this name is existing.

I feel more and more confuse by documentation and example code that seems to be outdated and/or not complete.

Please, can someone help to provide real documentation or example code about how to /* Do Something... */ with librpm API in C code ?

**Comment by rpluem-vf** [comment]
Try calling `rpmReadConfigFiles(NULL, NULL);` first to have the library properly initialized from the configuration files in the default locations.

**Comment by 3millionminds** [comment]
For anyone else looking for an implementation for 6.x, check out this post/thread: https://github.com/rpm-software-management/rpm/discussions/4168#discussioncomment-16379080

**Comment by dralley** [comment]
Example using librpm.rs (the Rust bindings):

```Rust
use librpm::{Db, Index};

fn main() {
    librpm::init().expect("failed to initialize librpm");

    let db = Db::open().expect("failed to open RPM database");

    // Look up a package by name (exact match)
    for pkg in db.find(Index::Name, "rpm") {
        println!("{}", pkg.nevra());
    }
}
```
